### PR TITLE
Add optional node container target to Portal

### PR DESCRIPTION
**Pull request by Warsaken**
This PR aims to solve the issue #666.

It introduces a new optional prop for Portal: `nodeContainer` that allows the user to specify the containing node to render the Portal in.

If this PR is approved, then we can create a follow up PR to pass this prop to other relevant packages.

---

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code (Compile and run).
- [ ] Add or edit tests to reflect the change (Run with `yarn test`).
- [ ] Add or edit Storybook examples to reflect the change (Run with `yarn start`).
- [x] Ensure formatting is consistent with the project's Prettier configuration.
- [x] Add documentation to support any new features.

This pull request:

- [ ] Creates a new package
- [ ] Fixes a bug in an existing package
- [x] Adds additional features/functionality to an existing package
- [ ] Updates documentation or example code
- [ ] Other

### CODE REVIEW

File: packages/portal/src/index.tsx [hunk 1]
```diff
@@ -20,7 +20,11 @@ import { createPortal } from "react-dom";
  *
  * @see Docs https://reach.tech/portal#portal
  */
-const Portal: React.FC<PortalProps> = ({ children, type = "reach-portal" }) => {
+const Portal: React.FC<PortalProps> = ({
+  children,
+  type = "reach-portal",
+  nodeContainer,
```
Comment (chaance) [inline packages/portal/src/index.tsx:26]: Can we rename this prop to simply `container`?

File: packages/portal/src/index.tsx [hunk 2]
```diff
@@ -32,11 +36,12 @@ const Portal: React.FC<PortalProps> = ({ children, type = "reach-portal" }) => {
     // In that case, it's important to append to the correct document element.
     const ownerDocument = mountNode.current!.ownerDocument;
     portalNode.current = ownerDocument?.createElement(type)!;
-    ownerDocument!.body.appendChild(portalNode.current);
+    const container = nodeContainer || portalNode.current!.ownerDocument.body;
```
Comment (chaance) [inline packages/portal/src/index.tsx:39]: I think we can simplify this a tiny bit

```
let body = container || ownerDocument.body
body.appendChild(portalNode.current)
```

File: packages/portal/src/index.tsx [hunk 3]
```diff
@@ -64,6 +69,12 @@ type PortalProps = {
    * @see Docs https://reach.tech/portal#portal-type
    */
   type?: string;
+  /**
+   * Target DOM element to render the portal.
+   *
+   * @see Docs https://reach.tech/portal#portal-nodeContainer
+   */
+  nodeContainer?: HTMLElement;
```
Comment (chaance) [inline packages/portal/src/index.tsx:77]: Let's make `container` of type `Node`, `Document` is the default and really is can be any Node type.

File: website/src/pages/portal.mdx [hunk 4]
```diff
@@ -62,9 +62,11 @@ Renders content inside of a portal at the end of the DOM tree.
...
 #### Portal Props
 
-| Prop                    | Type | Required |
-| ----------------------- | ---- | -------- |
-| [`children`](#children) | node | true     |
+| Prop                              | Type        | Required |
+| --------------------------------- | ----------- | -------- |
+| [`children`](#children)           | node        | true     |
+| [`type`](#type)                   | string      | false    |
+| [`nodeContainer`](#nodeContainer) | HTMLElement | false    |
```
Comment (chaance) [inline website/src/pages/portal.mdx:69]: Update the prop name and type here as well!

File: website/src/pages/portal.mdx [hunk 5]
```diff
@@ -83,3 +85,9 @@ Any content you want to render inside of the portal.
 _Type_: `string` default: `reach-portal`
 
 The DOM element type to render.
+
+##### Portal nodeContainer
+
+_nodeContainer_: `HTMLElement` default: `document.body`
+
+The DOM element container of the portal. If not set the portal will be appended to the body of the document.
```
Comment (chaance) [inline website/src/pages/portal.mdx:93]: Update the prop name and type here as well, and tweak to read "The DOM **node** container..."

**Review by chaance — CHANGES_REQUESTED** [review]
Thanks for the PR! A few changes requested but I'd like to merge this when we can.

**Comment by Warsaken** [comment]
@chaance thanks for reviewing, I made the requested changes now.

**Comment by chaance** [comment]
On second thought I think it makes a lot more sense for this prop to accept a ref rather than a DOM node. Probably requested changes before I thought about it more fully, so sorry about that. Let's hold for right now, I'll circle back to this after I've taken care of a few high-priority bugs for the next release.

**Comment by Warsaken** [comment]
>  On second thought I think it makes a lot more sense for this prop to accept a ref rather than a DOM node. Probably requested changes before I thought about it more fully, so sorry about that. Let's hold for right now, I'll circle back to this after I've taken care of a few high-priority bugs for the next release.

Several of the other libraries I encountered write it as an `HTMLElement` (e.g.: https://github.com/palantir/blueprint/blob/develop/packages/core/src/components/portal/portal.tsx) that's why I started with that type, since it's familiar and easy to migrate existing code to Reach later. But let me know which type you think is the best here and I can make those changes.

**Comment by Warsaken** [comment]
@chaance This has been here for a while so I went ahead and updated the type to a `ref` like you suggested.

Being of `node` type or `HTMLElement` also seemed reasonable to me, for example in my use case the container I need to append the portal to is obtained through a function that returns a `HTMLElement` which may not even be a react component.
But either way works, I just need to convert that to the ref format afterwards.

Let me know if there are any other changes necessary. Thanks.

**Comment by zefj** [comment]
hey @chaance , we could really use this feature too. Our case involves `menu-button` within overflow: hidden nodes living within modals, so unfortunately we cannot just disable the portal :disappointed: Is there any chance of getting this soon?

**Review by chaance — APPROVED** [review]

**Comment by Warsaken** [comment]
I see the PR was merged, thanks @chaance.

Now we need to create a PR to pass the props along for the other libraries that use Portal (e.g. Dialog).
Will you, or someone else, create a PR for that or should I do it?

**Pull request merged**